### PR TITLE
Transfer - Fix delayed artifacts mechanism

### PR DESCRIPTION
**Pull request by yahavi**
- [x] All [tests](https://github.com/jfrog/jfrog-cli-core#tests) passed. If this feature is not already covered by the tests, I added new tests.
- [x] All [static analysis checks](https://github.com/jfrog/jfrog-cli-core/actions/workflows/analysis.yml) passed.
- [x] This pull request is on the dev branch.
- [x] I used gofmt for formatting the code before submitting the pull request.
-----

Tests PR: https://github.com/jfrog/jfrog-cli/pull/2204

Enhance and optimize the following elements:
   - Delaying pom.xml files, not just `*.pom`.
   - Eliminating the need for recursion in delaying artifacts. Instead, transfer them in a sorted manner. This approach is significantly more efficient, especially considering that in Docker and Conan repositories, we triggered 2-3 invocations of doTransferWithProducerConsumer.

### CODE REVIEW

File: artifactory/commands/transferfiles/delayedartifactshandler.go [hunk 1]
```diff
@@ -168,7 +160,8 @@ func consumeDelayedArtifactsFiles(pcWrapper *producerConsumerWrapper, filesToCon
 			return err
 		}
 
-		shouldStop, err := uploadByChunks(delayedArtifactsFile.DelayedArtifacts, uploadChunkChan, base, delayHelper, errorsChannelMng, pcWrapper)
+		sortedDelayedArtifacts := sortDelayedArtifacts(delayedArtifactsFile, delayHelper)
```
Comment (RobiNino) [inline artifactory/commands/transferfiles/delayedartifactshandler.go:163]: I believe this is only sorting the artifacts within a specific delayedArtifactsFile. Those files are handled in parallel, meaning that even if a single file is being handled in the correct order, artifacts from different files can be uploaded in the wrong order.

File: artifactory/commands/transferfiles/delayedartifactshandler.go [hunk 2]
```diff
@@ -246,8 +255,10 @@ type delayUploadHelper struct {
 }
 
 // Decide whether to delay the deployment of a file by running over the shouldDelayUpload array.
-// When there are multiple levels of requirements in the deployment order, the first comparison function in the array can be removed each time in order to no longer delay by that rule.
 func (delayHelper delayUploadHelper) delayUploadIfNecessary(phase phaseBase, file api.FileRepresentation) (delayed, stopped bool) {
+	if phase.phaseId != api.Phase1 {
```
Comment (RobiNino) [inline artifactory/commands/transferfiles/delayedartifactshandler.go:259]: Shouldn't files be delayed in other phases? For example:
Phase 2 - Found new docker image that contains a `manifest.json`.
Phase 3 - Handling failures, `manifest.json` can be one of them.

**Review by RobiNino — CHANGES_REQUESTED** [review]

File: artifactory/commands/transferfiles/delayedartifactshandler.go (hunk 1 above)
Comment (yahavi) [reply artifactory/commands/transferfiles/delayedartifactshandler.go]: Good catch! Thanks

File: artifactory/commands/transferfiles/delayedartifactshandler.go (hunk 2 above)
Comment (yahavi) [reply artifactory/commands/transferfiles/delayedartifactshandler.go]: You are right. Thanks

File: artifactory/commands/transferfiles/delayedartifactshandler.go [hunk 3]
```diff
@@ -149,32 +149,38 @@ func handleDelayedArtifactsFiles(filesToConsume []string, base phaseBase, delayU
...
-		shouldStop, err := uploadByChunks(delayedArtifactsFile.DelayedArtifacts, uploadChunkChan, base, delayHelper, errorsChannelMng, pcWrapper)
+		allDelayedArtifacts = append(allDelayedArtifacts, delayedArtifactsFile.DelayedArtifacts...)
+	}
+
+	// Upload delayed artifacts, if any exist
+	if len(allDelayedArtifacts) > 0 {
+		log.Info(fmt.Sprintf("Uploading %d delayed artifacts...", len(allDelayedArtifacts)))
+		allDelayedArtifacts = sortDelayedArtifacts(&allDelayedArtifacts, delayHelper)
+		// Use an empty delayUploadHelper to prevent further delays of these files.
+		shouldStop, err := uploadByChunks(allDelayedArtifacts, uploadChunkChan, base, delayUploadHelper{}, errorsChannelMng, pcWrapper)
```
Comment (attiasas) [inline artifactory/commands/transferfiles/delayedartifactshandler.go:174]: If a delayed artifact fails here, for example, 'conanfile.py', then the following file at the ordered delayed list is for example 'conaninfo.txt' and no issue at upload, won't it be a problem later when we try to upload It again?

**Pull request closed without merging**